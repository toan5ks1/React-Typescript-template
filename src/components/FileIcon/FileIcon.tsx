import { useMemo, HTMLAttributes, useCallback } from "react";

import { AudioType, ImageType, MIMETypes, VideoType } from "src/common/constants/enum";
import { convertString } from "src/common/constants/helper";
import { getFileExtension, isImage, isPdf, isVideo } from "src/common/utils/file";
import { $enum } from "ts-enum-util";

import { Box } from "@mui/material";
import AudioIcon from "src/components/SvgIcons/AudioIcon";
import CSVIcon from "src/components/SvgIcons/CSVIcon";
import FileIconImg from "src/components/SvgIcons/FileIcon";
import ImageIcon from "src/components/SvgIcons/ImageIcon";
import PDFIcon from "src/components/SvgIcons/PDFIcon";
import VideoIcon from "src/components/SvgIcons/VideoIcon";

type EnhanceMimeTypes = MIMETypes | string | undefined;

interface FileIconProps extends HTMLAttributes<HTMLImageElement> {
    type?: EnhanceMimeTypes;
    name?: string;
    className?: string;
}
const images = $enum(ImageType)
    .getValues()
    .map((e) => String(e).replace("image/", ""));

const videos = $enum(VideoType)
    .getValues()
    .map((e) => String(e).replace("video/", ""));
const pdfs = ["pdf"];
const csv = ["csv"];
const audios = $enum(AudioType)
    .getValues()
    .map((e) => String(e).replace("audio/", ""));

const FileIcon = ({ type, name, className, ...props }: FileIconProps) => {
    const getPath = useCallback((fileType: EnhanceMimeTypes, name?: string) => {
        const ext = getFileExtension(convertString(name));

        switch (true) {
            case csv.includes(ext):
                return <CSVIcon data-testid="FileIcon__CSV" />;
            case audios.includes(ext):
                return <AudioIcon data-testid="FileIcon__audio" />;
            case isImage({ type: fileType }):
            case images.includes(ext):
                return <ImageIcon data-testid="FileIcon__image" />;
            case isPdf({ type: fileType }):
            case pdfs.includes(ext):
                return <PDFIcon data-testid="FileIcon__PDF" />;
            case isVideo({ type: fileType }):
            case videos.includes(ext):
                return <VideoIcon data-testid="FileIcon__video" />;
            default:
                return <FileIconImg data-testid="FileIcon__file" />;
        }
    }, []);

    const image = useMemo(() => getPath(type, name), [getPath, name, type]);

    return (
        <Box
            className={className}
            sx={{
                minWidth: 28,
                minHeight: 28,
                display: "flex",
            }}
            data-testid="FileIcon"
            {...props}
        >
            {image}
        </Box>
    );
};

export default FileIcon;
