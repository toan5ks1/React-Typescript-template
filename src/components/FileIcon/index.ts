export * from "./FileIcon";
export { default } from "./FileIcon";
