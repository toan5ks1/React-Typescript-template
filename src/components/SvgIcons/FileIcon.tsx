import { HTMLAttributes } from "react";

const FileIcon = (props: HTMLAttributes<SVGElement>) => {
    return (
        <svg
            xmlns="http://www.w3.org/2000/svg"
            width={28}
            height={28}
            xmlSpace="preserve"
            {...props}
        >
            <image
                width="100%"
                height="100%"
                href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACQAAAAkCAQAAABLCVATAAAABGdBTUEAALGPC/xhBQAAACBjSFJN AAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAAAmJLR0QA/4ePzL8AAAAHdElN RQfmBBMPOinXog2mAAAA30lEQVRIx2NgoBJghLNEGOwZdIjQ0YhfWofhNcN/oiABMJ9IY3AaxAKl IxgYGBgYGigPo/9oIUYyYCJfK3avoQMHBnssojcYVhIyEDkgTRgu4wzqfFIMOo3DkN/EGIVsEC7X BBNjFDEGMRBjFHEGEWEUsQYRNIp4gwgYRYpBeI0izSCEUeEwAXKzyFqGCIY/DAwMGjABFhI04y2L qJZpRw2io0Gw6G+glkGNFJlCTa+NMIP+kKD/Cz6DNpBg0FF8kjYM34lsmUxHaMLebNBh8GbgIOCW HwwrGR6QE5oEAADy5/IIXX8MJwAAACV0RVh0ZGF0ZTpjcmVhdGUAMjAyMi0wNC0xOVQxMjo1ODo0 MSswMzowMD4idSYAAAAldEVYdGRhdGU6bW9kaWZ5ADIwMjItMDQtMTlUMTI6NTg6NDErMDM6MDBP f82aAAAAAElFTkSuQmCC"
            />
        </svg>
    );
};

export default FileIcon;
