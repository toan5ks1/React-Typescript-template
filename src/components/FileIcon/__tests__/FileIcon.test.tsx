import FileIcon from "../FileIcon";

import { render, screen } from "@testing-library/react";

describe("<FileIcon />", () => {
    it("should match snapshot", () => {
        const { container } = render(<FileIcon type={undefined} />);
        expect(container).toMatchSnapshot();
    });

    it("should render default when undefined", () => {
        render(<FileIcon type={undefined} />);
        expect(screen.getByTestId("FileIcon__file")).toBeInTheDocument();
    });
});
