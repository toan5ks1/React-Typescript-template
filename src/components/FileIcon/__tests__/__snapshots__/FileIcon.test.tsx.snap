// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileIcon /> should match snapshot 1`] = `
.emotion-0 {
  min-width: 28px;
  min-height: 28px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="MuiBox-root emotion-0"
    data-testid="FileIcon"
  >
    <svg
      data-testid="FileIcon__file"
      height="28"
      width="28"
      xml:space="preserve"
      xmlns="http://www.w3.org/2000/svg"
    >
      <image
        height="100%"
        href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACQAAAAkCAQAAABLCVATAAAABGdBTUEAALGPC/xhBQAAACBjSFJN AAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAAAmJLR0QA/4ePzL8AAAAHdElN RQfmBBMPOinXog2mAAAA30lEQVRIx2NgoBJghLNEGOwZdIjQ0YhfWofhNcN/oiABMJ9IY3AaxAKl IxgYGBgYGigPo/9oIUYyYCJfK3avoQMHBnssojcYVhIyEDkgTRgu4wzqfFIMOo3DkN/EGIVsEC7X BBNjFDEGMRBjFHEGEWEUsQYRNIp4gwgYRYpBeI0izSCEUeEwAXKzyFqGCIY/DAwMGjABFhI04y2L qJZpRw2io0Gw6G+glkGNFJlCTa+NMIP+kKD/Cz6DNpBg0FF8kjYM34lsmUxHaMLebNBh8GbgIOCW HwwrGR6QE5oEAADy5/IIXX8MJwAAACV0RVh0ZGF0ZTpjcmVhdGUAMjAyMi0wNC0xOVQxMjo1ODo0 MSswMzowMD4idSYAAAAldEVYdGRhdGU6bW9kaWZ5ADIwMjItMDQtMTlUMTI6NTg6NDErMDM6MDBP f82aAAAAAElFTkSuQmCC"
        width="100%"
      />
    </svg>
  </div>
</div>
`;
